module/stats_samtools.nf

```diff
@@ -12,29 +13,49 @@ process run_stats_SAMtools {
     publishDir path: "${params.workflow_output_dir}/output",
         pattern: "*stats.txt",
         mode: "copy",
-        enabled: true
+        enabled: true,
+        saveAs: { "${outdir}/${file(it).getName()}" }
 
     publishDir path: "${params.workflow_log_output_dir}",
         pattern: ".command.*",
         mode: "copy",
-        saveAs: { "${task.process.replace(':', '/')}-${id}/log${file(it).getName()}" }
+        saveAs: { "${task.process.replace(':', '/')}-${log_suffix}/log${file(it).getName()}" }
 
     input:
-        tuple val(orig_id), val(id), path(path), val(read_length), val(sample_type)
+        tuple path(path), val(unused), val(sm_id), val(rg_arg), val(rg_id), val(lib_id), val(unused), val(unused)
 
     output:
         path "*stats.txt"
         path ".command.*"
 
     script:
-    output_filename = generate_standard_filename("SAMtools-${params.samtools_version}",
-        params.dataset_id,
-        id,
-        [:])
+    if (task.process == "run_stats_SAMtools_sample") {
+        output_filename = generate_standard_filename("SAMtools-${params.samtools_version}",
+            params.dataset_id,
+            sm_id,
+            [:])
+        outdir = "."
+        log_suffix = "${sm_id}"
+    } else if (task.process == "run_stats_SAMtools_library") {
+        output_filename = generate_standard_filename("SAMtools-${params.samtools_version}",
+            params.dataset_id,
+            lib_id,
+            [:])
+        outdir = sm_id
+        log_suffix = "${sm_id}/${lib_id}"
+    } else if (task.process == "run_stats_SAMtools_readgroup") {
+        output_filename = generate_standard_filename("SAMtools-${params.samtools_version}",
+            params.dataset_id,
+            rg_id,
+            [:])
+        outdir = "${sm_id}/${lib_id}"
+        log_suffix = "${sm_id}/${lib_id}/${rg_id}"
+    }
+
```

Comment:
**suggestion:** Consider re-organizing this block to something like:
```Nextflow
if (params.stat_mode == 'library') {
    filename_id = lib_id
    outdir = ...
    log_suffix = ...
} else if (params.stat_mode == 'readgroup') {
    filename_id = rg_id
    outdir = ...
    log_suffix = ...
} else {
    filename_id = sm_id
    outdir = ...
    log_suffix = ...
}

output_filename = generate_standard_filename("SAMtools-${params.samtools_version}",
            params.dataset_id,
            filename_id,
            [:])
```
and then using `addParams` when importing the process to set `stat_mode`.

This reduces the amount of duplicated code (ex. the standard filename generation function is only coded once) and also removes the need for the process itself to have to be imported with a specific name and instead be controlled by a parameter set during import.